go-controller/pkg/ovn/base_network_controller_pods.go

```diff
@@ -611,138 +491,15 @@ func (bnc *BaseNetworkController) addLogicalPortToNetwork(pod *kapi.Pod, nadName
-					podDesc, switchName, err)
-			}
-		}
-		needsNewMacOrIPAllocation := false
-
-		// ensure we have reserved the IPs found in OVN
-		if len(podIfAddrs) == 0 {
-			needsNewMacOrIPAllocation = true
-		} else if bnc.doesNetworkRequireIPAM() {
-			if err = bnc.lsManager.AllocateIPs(switchName, podIfAddrs); err != nil && err != ipallocator.ErrAllocated {
-				klog.Warningf("Unable to allocate IPs %s found on existing OVN port: %s, for pod %s on switch: %s"+
-					" error: %v", util.JoinIPNetIPs(podIfAddrs, " "), portName, podDesc, switchName, err)
-
-				needsNewMacOrIPAllocation = true
-			}
-		}
-		if needsNewMacOrIPAllocation {
-			if network != nil && network.IPRequest != nil && !bnc.doesNetworkRequireIPAM() {
-				klog.V(5).Infof("Will use static IP addresses for pod %s on a flatL2 topology without subnet defined", podDesc)
-				podIfAddrs, err = calculateStaticIPs(podDesc, network.IPRequest)
-				if err != nil {
-					return nil, nil, nil, false, err
-				}
-				podMac = util.IPAddrToHWAddr(podIfAddrs[0].IP)
-			} else {
-				// Previous attempts to use already configured IPs failed, need to assign new
-				generatedPodMac, generatedPodIfAddrs, err := bnc.assignPodAddresses(switchName)
-				if err != nil {
-					return nil, nil, nil, false, fmt.Errorf("failed to assign pod addresses for pod %s on switch: %s, err: %v",
-						podDesc, switchName, err)
-				}
-				if podMac == nil {
-					podMac = generatedPodMac
-				}
-				if len(generatedPodIfAddrs) > 0 {
-					podIfAddrs = generatedPodIfAddrs
-				}
-			}
-		}
-
-		releaseIPs = true
-		// handle error cases separately first to ensure binding to err, otherwise the
-		// defer will fail
-		if network != nil && network.MacRequest != "" {
-			podMac, err = calculateStaticMAC(podDesc, network.MacRequest)
-			if err != nil {
-				return nil, nil, nil, false, err
-			}
-		}
-		podAnnotation = &util.PodAnnotation{
-			IPs: podIfAddrs,
-			MAC: podMac,
-		}
-		var nodeSubnets []*net.IPNet
-		if nodeSubnets = bnc.lsManager.GetSwitchSubnets(switchName); nodeSubnets == nil && bnc.doesNetworkRequireIPAM() {
-			return nil, nil, nil, false, fmt.Errorf("cannot retrieve subnet for assigning gateway routes for pod %s, switch: %s",
-				podDesc, switchName)
-		}
-		err = bnc.addRoutesGatewayIP(pod, network, podAnnotation, nodeSubnets)
-		if err != nil {
-			return nil, nil, nil, false, err
-		}
-
-		klog.V(5).Infof("Annotation values: ip=%v ; mac=%s ; gw=%s",
-			podIfAddrs, podMac, podAnnotation.Gateways)
-		annoStart := time.Now()
-		err = bnc.updatePodAnnotationWithRetry(pod, podAnnotation, nadName)
-		podAnnoTime := time.Since(annoStart)
-		klog.Infof("[%s] addLogicalPort annotation time took %v", podDesc, podAnnoTime)
-		if err != nil {
-			return nil, nil, nil, false, err
-		}
-		releaseIPs = false
+	podAnnotation, annotationUpdated, err := bnc.allocateIPs(pod, existingLSP, nadName, network)
+	if err != nil {
```

Comment:
so these changes are very scary imo..it took us a long time to get the addLogicalPort path solid, and I'm pretty hesitant to make any change here. I would prefer you just change the path for localnet and leave primary network alone.

Reply:
OK, I can do that. 

But I have thoughts about this.

If we cannot refactor this code, then that makes the code unmaintainable from my perspective. I respected all the comments the code had and did not brake any tests. In exchange, I added more comments and tests. Yes, there might be regressions for things which we had no comments or no tests, but if we ever want to improve this, we need to take the plunge.

Reply:
I agree that we should be able to refactor it at some point, I just dont think the time is now with all of the other risks and code change related to IC. When we had issues with the addLogicalPort stuff in the past it was very complicated to debug and go back in time and figure out exactly what happened to cause the problems. I dont want to risk going down that path while trying to test out our new architecture in downstream. I would vote we have branches for now (one primary one everything else) and then after we get past stabilizing IC, we consolidate to the 1 path.

That will also give us some time to make sure the changes you make work well for secondary networks before we try it on primary.

Reply:
I moved the old allocation for the default network to `allocatePodAnnotation` and move the new allocation for secondary networks to `allocatePodAnnotationForSecondaryNetwork`.

Keeping the old functionality in the original addLogicalPortToNetwork would have just been messy, so I tried to reach you midway.

The old functionality was moved intact. The only differences are
-  I removed `lsp.DynamicAddresses = nil`, we were not writing that as `DynamicAddresses` was not in the list of updated fields we pass to libovsdb client.
- `AddRoutesGatewayIP` still lives in `util/pod_annotation.go` so that it can be reused. I carries a change I introduced to derive the subnets from the pod IPs but is otherwise intact.